contributors/design-proposals/storage/volume-snapshotting.md

```diff
@@ -21,498 +25,529 @@ For the first version of snapshotting support in Kubernetes, only on-demand snap
+        // Status of the condition, one of True, False, Unknown.
+        Status core_v1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
+        // The last time the condition transitioned from one status to another.
+        // +optional
+        LastTransitionTime metav1.Time `json:"lastTransitionTime" protobuf:"bytes,3,opt,name=lastTransitionTime"`
+        // The reason for the condition's last transition.
+        // +optional
+        Reason string `json:"reason" protobuf:"bytes,4,opt,name=reason"`
+        // A human readable message indicating details about the transition.
+        // +optional
+        Message string `json:"message" protobuf:"bytes,5,opt,name=message"`
+}
+
+// Represents the actual location and type of the snapshot. Only one of its members may be specified.
+type VolumeSnapshotDataSource struct {
+	// HostPath represents a directory on the host.
+	// Provisioned by a developer or tester.
+	// This is useful for single-node development and testing only!
+	// On-host storage is not supported in any way and WILL NOT WORK in a multi-node cluster.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#hostpath
+	// +optional
+	HostPath *HostPathVolumeSnapshotSource `json:"hostPath,omitempty"`
+	// AWSElasticBlockStore represents an AWS Disk resource that is attached to a
+	// kubelet's host machine and then exposed to the pod.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#awselasticblockstore
+	// +optional
+	AWSElasticBlockStore *AWSElasticBlockStoreVolumeSnapshotSource `json:"awsElasticBlockStore,omitempty"`
+
+    // ... etc. for other snapshot types
+}
+```
+
+An example of the `VolumeSnapshotDataSource` for Amazon EBS snapshots:
+
+```
+// AWS EBS volume snapshot source
+type AWSElasticBlockStoreVolumeSnapshotSource struct {
+	// Unique id of the persistent disk snapshot resource. Used to identify the disk snapshot in AWS
+	SnapshotID string `json:"snapshotId"`
+}
+```
+
+### Event loop in controller
+
+The volume snapshot controller maintains two data structures (ActualStateOfWorld and DesiredStateOfWorld) and
+periodically reconciles the two. The data structures are being update by the API sever event handlers.
+
+* If a new `VolumeSnapshot` is added, the *add* handler adds it to the DesiredStateOfWorld (DSW)
+
+* If a `VolumeSnapshot` is deleted, the *delete* handler removes it from the DSW.
+
+* Reconciliation loop in the controller
+
+    * For every `VolumeSnapshot` in the ActualStateOfWorld (ASW) find the corresponding `VolumeSnapshot` in the DSW.
+    If such a snapshot does not exist, start a snapshot deletion operation:
+
+        * Determine the correct volume snapshot plugin to use from the `VolumeSnapshotData` referenced by the
+        `VolumeSnapshot`
+
+        * Create a delete operation: only one such operation is allowed to exist for the given `VolumeSnapshot` and
```

Comment:
If the controller stopped before we delete `VolumeSnapshotData`, according `VolumeSnapshot` no longer exists in `ASW` after controller started again, and the `VolumeSnapshotData` will be orphaned. Maybe we need an asynchronous go routine to clean up these `VolumeSnapshotDatas`.